### PR TITLE
Unify memory implementation under single trait, fix Linux implementation

### PR DESCRIPTION
**Pull request by paskal**
Also, run CI not only in master branch.

### CODE REVIEW

File: src/memory.rs [hunk 1]
```diff
@@ -3,47 +3,62 @@ use std::mem::size_of;
...
 
-    pub fn search<T: Integer>(&mut self, value: T) {
-        self.candidates = os_specific::search_everywhere(self.pid, value);
+    pub fn search(&mut self, value: T) {
+        self.candidates = self.os.search_everywhere(self.pid, value);
     }
 
-    // FIXME: If refine is called with a differnet larger type, it can try access memory outside the process memory map
-    pub fn refine<T: Integer>(&mut self, new_value: T) {
+    // FIXME: If refine is called with a different larger type, it can try access memory outside the process memory map
```
Comment (sanchopanca) [inline src/memory.rs]: After you changes this is no longer a concern

File: src/memory/windows.rs [hunk 2]
```diff
@@ -10,125 +10,136 @@ use winapi::um::winnt::{
...
+
+        for &address in candidates {
+            let mut bytes_read = 0;
+            let mut value_at_address = T::new();
+            let read_success = unsafe {
+                ReadProcessMemory(
+                    process_handle,
+                    address as *const _,
+                    &mut value_at_address as *mut _ as *mut _,
+                    size_of::<i32>(),
```
Comment (sanchopanca) [inline src/memory/windows.rs]: This is not related to your PR, but I think this is a bug. It should be `size_of::<T>`. You can fix it, or I can fix it later.

File: tests/test_on_lab_rat.rs [hunk 3]
```diff
@@ -53,6 +56,9 @@ fn test_modify_lab_rat_memory() {
         .expect("Failed to read line from stdout");
     print!("Lab Rat says: {}", line);
 
+    // Add a delay here
+    thread::sleep(Duration::from_secs(1));
```
Comment (sanchopanca) [inline tests/test_on_lab_rat.rs]: Was it not working without the delay? Synchronization via sleep often leads to flaky tests

File: Cargo.toml [hunk 4]
```diff
@@ -32,4 +32,5 @@ winapi = { version = "0.3.9", features = [
     "processenv",
     "processthreadsapi",
     "memoryapi",
+    "winnt"
```
Comment (sanchopanca) [inline Cargo.toml:35]: What functionality reququires this feaure? Maybe I missed something during the review.

**Review by sanchopanca — COMMENTED** [review]
Thanks for looking into this! Left some comments

BTW, did you run `cargo fmt` and `cargo clippy`?

File: Cargo.toml (hunk 4 above)
Comment (paskal) [reply Cargo.toml:35]:
```
use winapi::um::winnt::{
    MEMORY_BASIC_INFORMATION, PROCESS_QUERY_INFORMATION, PROCESS_VM_OPERATION, PROCESS_VM_READ,
    PROCESS_VM_WRITE,
};
```

File: tests/test_on_lab_rat.rs (hunk 3 above)
Comment (paskal) [reply tests/test_on_lab_rat.rs]: Added while troubleshooting, not needed. Removed it.

File: src/memory/windows.rs (hunk 2 above)
Comment (paskal) [reply src/memory/windows.rs]: Fixed.

File: src/memory.rs (hunk 1 above)
Comment (paskal) [reply src/memory.rs]: Removed.

**Pull request merged**